File: vane-portals/src/main/java/org/oddlama/vane/portals/event/PortalChangeSettingsEvent.java
package org.oddlama.vane.portals.event;

import org.bukkit.entity.Player;
import org.bukkit.event.HandlerList;
import org.oddlama.vane.portals.portal.Portal;

public class PortalChangeSettingsEvent extends PortalEvent {

    private static final HandlerList handlers = new HandlerList();
    private Player player;
    private Portal portal;
    private boolean checkOnly;
    private boolean cancelIfNotOwner = true;

    public PortalChangeSettingsEvent(final Player player, final Portal portal, boolean checkOnly) {
        this.player = player;
        this.portal = portal;
        this.checkOnly = checkOnly;
    }

    public void setCancelIfNotOwner(boolean cancelIfNotOwner) {
        this.cancelIfNotOwner = cancelIfNotOwner;
    }

    public Player getPlayer() {
        return player;
    }

    public Portal getPortal() {
        return portal;
    }

    public boolean checkOnly() {
        return checkOnly;
    }

    public HandlerList getHandlers() {
        return handlers;
    }

    public static HandlerList getHandlerList() {
        return handlers;
    }

    @Override
    public boolean isCancelled() {
        var cancelled = super.isCancelled();
        if (cancelIfNotOwner) {
            cancelled |= !player.getUniqueId().equals(portal.owner());
        }
        return cancelled;
    }
}

File: vane-portals/src/main/kotlin/org/oddlama/vane/portals/event/PortalChangeSettingsEvent.kt
package org.oddlama.vane.portals.event

import org.bukkit.entity.Player
import org.bukkit.event.HandlerList
import org.oddlama.vane.portals.portal.Portal

/**
 * Fired while validating or applying portal settings changes.
 *
 * @property player the acting player.
 * @property portal the portal being modified.
 * @property checkOnly whether this is a dry-run validation.
 */
class PortalChangeSettingsEvent(@JvmField val player: Player, @JvmField val portal: Portal, private val checkOnly: Boolean) :
    PortalEvent() {
    /** Controls whether non-owners are automatically denied. */
    private var cancelIfNotOwner = true

    /** Sets whether settings changes are automatically cancelled for non-owners. */
    fun setCancelIfNotOwner(cancelIfNotOwner: Boolean) {
        this.cancelIfNotOwner = cancelIfNotOwner
    }

    /** Returns whether this event is in validation-only mode. */
    fun checkOnly() = checkOnly

    /** Returns Bukkit handlers for this event. */
    override fun getHandlers() = handlerList

    /** Returns true when explicitly cancelled or blocked by ownership checks. */
    override fun isCancelled() = super.isCancelled() || (cancelIfNotOwner && player.uniqueId != portal.owner())

    /** Static Bukkit handler list for this event type. */
    companion object {
        /** Shared handler list used by Bukkit's event system. */
        @JvmStatic val handlerList: HandlerList = HandlerList()
    }
}
